spec.bs

```diff
@@ -1068,34 +1260,79 @@ Note: Two {{PAHistogramContribution}}s for the same report can be merged if and
-ignore>api</var> and a [=moment=] <var ignore>currentTime</var> and a
-[=boolean=] |consumeIfPermitted|, perform the following steps. They return a
-[=list=] of [=booleans=] with the same [=list/size=] as |contributions|.
+|contributions|, a [=list=] of [=named budget ledgers=]
+|remainingBudgetLedgers|, an [=origin=] |origin|, a [=context type=] |api|, a
+[=moment=] |currentTime|, and a [=boolean=] |consumeIfPermitted|, perform the
+following steps. They return a [=list=] of [=booleans=] with the same
+[=list/size=] as |contributions|.
+
+1. If |remainingBudgetLedgers| is [=list/empty=]:
+    1. [=list/For each=] |budgetingScope| of [=all budgeting scopes=]:
+        1. Let |remainingBudgetLedger| be a new [=named budget ledger=].
+        1. [=map/Set=] |remainingBudgetLedger|[null] to |budgetingScope|'s
+            [=budgeting scope/max budget=].
+        1. [=list/Append=] |remainingBudgetLedger| to |remainingBudgetLedgers|.
+1. Let |numScopes| be the [=list/size=] of [=all budgeting scopes=].
+1. [=Assert=] that |numScopes| is equal to the [=list/size=] of
+    |remainingBudgetLedgers|.
+1. [=list/For each=] |i| in [=the exclusive range|the range=] from 0 to
+    |numScopes|, exclusive:
+    1. Let |budgetingScope| be [=all budgeting scopes=][|i|].
+    1. Let |duration| be |budgetingScope|'s [=budgeting scope/duration=].
+    1. Let |remainingBudgetLedger| be |remainingBudgetLedgers|[|i|].
+    1. [=map/For each=] |name| → |maxBudget| of |remainingBudgetLedger|:
+        1. Let |budgetConsumed| be an integer determined by an
+            [=implementation-defined=] algorithm given |origin|, |api|, |name|,
+            and |duration|. This algorithm should return a value greater than or
+            equal to the sum of the values of contributions that referenced the
+            named budget |name| and were accepted in the last |duration| units
+            of time. The contributions must have originated from an isolated
+            context of type |api| that was loaded from |origin|.
+        1. [=Assert=] that |maxBudget| is less than or equal to
+            |budgetingScope|'s [=budgeting scope/max budget=].
+        1. If |budgetConsumed| is greater than or equal to |maxBudget|:
+            1. [=map/Set=] |remainingBudgetLedger|[|name|] to 0.
+        1. Otherwise:
+            1. [=map/Set=] |remainingBudgetLedger|[|name|] to |maxBudget| -
+                |budgetConsumed|.
+
+    Note: At this point, we have adjusted the values of each ledger in
+    |remainingBudgetLedgers| based on |origin|'s historical budget consumption.
+    Next, we will accept contributions on a first-come, first-served basis, by
+    subtracting the value of each contribution from the ledgers in
+    |remainingBudgetLedgers| until one would become empty.
+
 1. Let |resultForEachContribution| be a new [=list=] of [=booleans=].
-1. Let |approvedValueSum| be 0.
+1. Let |ledgerConsumed| be a new [=named budget ledger=].
```

Comment:
This conflicts a bit with the named budget ledger definition above as "track[ing] the remaining budget". We might want to adjust the definition (or just maintain a map here).